### ISSUE TITLE
Long strip mode

### ISSUE THREAD

**Issue by scanlime** [scanlime/fadecandy#2]
Currently, the Fadecandy firmware uses a fixed strip length of 64 LEDs in order to support temporal dithering. Longer strips have a lower maximum update rate, and much above 64 LEDs there's too much flickering for it to work.

Strips with longer than 64 LEDs will never be the recommended configuration for WS2811 LEDs with Fadecandy, but it would be really useful to have a new firmware configuration mode that would support these long strips anyway.

Internally, this would require some refactoring so that we can use different RAM layouts for different firmware configuration modes. Most of the bits in the first byte of firmware configuration space will probably turn into a "mode" number, which selects a RAM layout and a redraw engine to use for that mode.

This would also require some fcserver support, so it knows about the different strip lengths. And any layout editor tool we build would really like to know this as well.

**Comment by rewolff** [comment]
each pixel reads the signal and outputs its own interpretation of that signal. So the last pixel after 30m of led strip will see the signal generated by the pixel before it, not a "degraded" version of what your controller sent out at the beginning of the strip. 

What IS important is that you have a proper 5V power supply. This may mean that you have to put a thick 5V wire along the strip and connect it to the strip every now and then. With a 64-led-per string limit, every 60 or 64  leds would be fine. Go for 3A per 60 leds. Ah, yes, that comes to 18A for 360 leds.

**Comment by SeanConnell** [comment]
TL;DR: **Being able to trade off temporal dithering against LED length would enable cost savings and keep fadecandy relevant in the face of increasing LED density.** 

I think this limitation is a big issue for fadecandy today.  LED strips have increased in density so that very few are in the 64 range anymore. To continue using off the shelf components it would be very nice to be able to disable temporal dithering and still be able to use the other niceties of the system (opc_server, 8 channels/density, frame interpolation, good update rate, etc). 

Are there any other high level reasons that we couldn't drive more leds per strip without temporal dithering at an acceptable rate? (I'm mostly thinking memory, but I'm not familiar with the code yet) A naive bitrate (800kbps as mentioned on datasheet) calculation against 256 leds shows `800000 / (24 * 256) = 130` we should be able to hit 120hz with overhead. I think this is a fair tradeoff for being able to drive 2056 LEDs, or at least significantly reduce cost for projects (larger displays are cheaper per led).

As a concrete motivating example, I've been playing around with some of the 16x16 ws2812b displays available. I hacked in 3 extra channels (mechanically/bodge wires) so I could drive it with the fade candy but it was tough enough to do that the display isn't reliable anymore. The alternative is buying 4 8x8 displays, but they're 4x the cost/led which is significant (16 for 1 8x8, 35 for 1 16x16). There are also very nicely packaged strips that would require similar channel hacks to drive. (I have one of these as well, at 170 leds) This is, pragmatically speaking, a big bummer and stumbling block for ease of use. I know "serious" artists do custom sizes/fabrication, but the easier the better for everybody.

I wanted to know if there was anything I should know before diving in to figure out an approach. Including such important details as...
- already fixed in BRANCH/OTHER REPO
- not an issue because project replaced by REPLACEMENT
- some other external factor/hack people use to get around this

Also any code pointers/gotchas would by nice too.

Thanks!

PS The fadecandy is a wonderful device. Thanks for making it.

**Comment by scanlime** [comment]
The biggest hurdle in any change to display format is in adding support for
multiple formats. That will be a lot of work in the firmware and fcserver.

Re longer strips with dithering, you will be out of cpu and ram quite soon,
but the frame rate is the biggest limit. With 800kbps you can’t add many
more leds before the dithering rates get too low for the effect to work.

-m

**Comment by SeanConnell** [comment]
I'm not worried about the multi format support (probably solved by a protocol handshake at startup or something) although perhaps I should be. At least this is mostly implementation complexity vs a hard limit. 

The 800kbps is more of an issue it sounds like. Is 400hz is really as low as one realistically wants to go then?

I guess perhaps the easier way out here then is to contact an ali manufacturer and try and get LED panels wired in groups of 64. That, and probably try and influence world semi to increase the bitrate of the WS281*s (no idea how to do the latter realistically, but I sent them an email).

**Comment by scanlime** [comment]
The "protocol handshake or something" is a big pandora's box here- if you
want to make a fork of the project that isn't compatible that's easier,
assuming you also swap in a larger processor. But doing this in a
compatible way involves adding some extensibility support to protocol
layers that aren't particularly flexible, like the USB protocol and
possibly OPC (for e.g. RGBW support). Also, there's no automatic firmware
update support, or support in fcserver for speaking different protocol
versions to different firmware versions. I don't want to discourage you, I
just want to make it clear that this is not a trivial piece of work. I also
want to reiterate that you'll need more RAM and CPU to control
significantly more LEDs; the dithering and double-buffering require more
RAM than a plain single framebuffer.

Regarding the dithering frame rate; 400hz is nice, you can go lower but it
starts to drastically limit how much extra dynamic range you can get, and
the flickering gets a lot worse. I chose to optimize for quality rather
than wiring, so that's how it's designed. If you wanted to try modulating
fewer longer strips at a slower rate, you'd need to change the firmware
somewhat significantly to rearrange the way the data is packed, and it
still may need a little bit more RAM than what you have. The 8 strip layout
is trying to make optimal use of the byte-wide DMA path, so any other
number will end up using the RAM less efficiently.

-m

**Comment by SeanConnell** [comment]
Not discouraged; I appreciate understanding the scope and all of the details are very helpful. I'm still trying to figure out the shape this project would take and overall feasibility. 

I'm trying to not trade off quality (there's plenty of "bad" LED driving, I've done my fair share...), but it'd be really nice to avoid all the headaches of wiring. And pragmatically, the reason I started looking into this, is that it's a lot more expensive to use N smaller displays. 

It sounds like fadecandy is very efficiently packed into existing hardware and it might be easier to take another route (beaglebone black) to go further. 

That being said, the main limitation remains that update rate. Fingers crossed that maybe world semi is planning on increasing that on the next rev of their drivers. 

Thanks for all the details/help in looking at feasibility; I'm not going to try and change anything, mostly based on the fact that regardless of how the fadecandy pushes the pixels, the WS2812's are actually the limit. Hopefully this exchange helps other people understand what the limitations are based on/the hurdles they'd need to clear/tradeoffs they'd need to make to tackle this.

**Comment by rewolff** [comment]
WorldSemi have an updated chip. Still 800kbps. 

The thing is, 800kbps is what you can still do with an arduino, and you're not playing with nanoseconds so that a nanosecond offset might  make a blt flip....

**Comment by manmeetjammu** [comment]
Ok Dears and especially @mica

If someone can drive ws2811 strips   
So long with arduino, why cannot you guys. I mean just drive with relatively lower  speeds . Please add this mode. See link below and joshs work .

https://wp.josh.com/2014/05/13/ws2812-neopixels-are-not-so-finicky-once-you-get-to-know-them/

And here is Tims work and a library link he created  for arduino is also there. Read carefully. You might get motivation from code. And how it works.
Please donot insult as i know nothing about coding and hardware.

https://cpldcpu.wordpress.com/2014/01/14/light_ws2812-library-v2-0-part-i-understanding-the-ws2812/

I Just want to point . Nothing else as few people just want to drive very long strips for show purpose. A few cheap controllers are there which can drive  very long strips like ws2811 but cannot do what fadecandy does. Like using processing and spout to talk to pixels.

**Comment by scanlime** [comment]
Fadecandy isn't designed to be a flexible tool that drives these LEDs in every possible mode, it's designed to do exactly one thing well. That's all.

**Comment by widestreid** [comment]
I would think that a strip with a length greater than 64 LEDs would be possible by putting each strip onto a different channel of the board and then physically aligning the strips to form one long strip. No? That way all the processing would be done through whatever was driving the fadecandy server. Could this be done by modifying the line of code "opc.ledStrip(0, 64, width/2, height/2, width / 70.0, 0, false);" in the following from the GitHub example to be >64? 

OPC opc;

PImage dot;

void setup()
{
  size(800, 200);

  // Load a sample image
  dot = loadImage("color-dot.png");

  // Connect to the local instance of fcserver
  opc = new OPC(this, "127.0.0.1", 7890);

  // Map one 64-LED strip to the center of the window
  opc.ledStrip(0, 64, width/2, height/2, width / 70.0, 0, false);
}
void draw()
{
  background(0);
 // Draw the image, centered at the mouse location
  float dotSize = width * 0.2;
  image(dot, mouseX - dotSize/2, mouseY - dotSize/2, dotSize, dotSize);
}

**Comment by widestreid** [comment]
Studied this in a lot more detail last night. I may have answered my own question because it seems that there is a way to setup the opc.ledGrid function something like this. 

opc.ledGrid(0, 2, width * 0.25, height/2, ledSpacing, stripSpacing, angle, zigzag, flip) 
opc.ledGrid(64, 2, width * 0.75, height/2, ledSpacing, stripSpacing, angle, zigzag, flip) 

Then assuming the processing window was  128 pixels wide, you would set the LED spacing to be 1 and the strip spacing to be 0. I'll have to test it later, thanks.
